File: src/main/java/net/jezevcik/workers/impl/AsynchronousWorker.java
package net.jezevcik.workers.impl;

import net.jezevcik.workers.Worker;
import net.jezevcik.workers.log.WorkerLoggerInterface;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * Runs all tasks in a different thread.
 */
public class AsynchronousWorker extends Worker {

    /**
     * The list of created threads.
     */
    private final List<Thread> threads = new ArrayList<>();

    public AsynchronousWorker(final WorkerLoggerInterface loggerInterface, final String name, final Object... waiting) {
        super(loggerInterface, name, waiting);
    }

    @Override
    public synchronized void run() {
        tasks.forEach(runnable -> {
            final Thread thread = new Thread(() -> {
                try {
                    runnable.run();
                } catch (Exception e) {
                    this.loggerInterface.accept(false, "Worker {} has crashed", e, this.name);
                    crashed = true;
                }
            });

            thread.start();

            threads.add(thread);
        });

        new Thread(() -> {
            while (!finished || crashed) {
                final Iterator<Thread> iterator = threads.iterator();

                if (!iterator.hasNext())
                    break;

                while (iterator.hasNext()) {
                    if (!iterator.next().isAlive())
                        iterator.remove();
                }
            }

            finished = true;

            this.loggerInterface.accept(false, "Worker {} finished!", null, this.name);

            for (Object o : waiting) {
                synchronized (o) {
                    o.notify();
                }
            }
        }).start();
    }

}

File: src/main/kotlin/net/jezevcik/workers/impl/AsynchronousWorker.kt
package net.jezevcik.workers.impl

import net.jezevcik.workers.log.WorkerLoggerInterface
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Runs all tasks in a different thread.
 */
class AsynchronousWorker(loggerInterface: WorkerLoggerInterface, name: String, vararg waiting: Any)
    : Worker(loggerInterface, name, *waiting) {

    /**
     * The list of created threads.
     */
    private val threads = CopyOnWriteArrayList<Thread>()

    override fun run() {
        tasks.forEach { runnable ->
            val thread = Thread {
                try {
                    runnable.run()
                } catch (e: Exception) {
                    loggerInterface.accept(false, "Worker {} has crashed", e, name)
                    crashed = true
                }
            }
            thread.start()
            threads.add(thread)
        }

        Thread {
            while (!finished || crashed) {
                val iterator = threads.iterator()
                if (!iterator.hasNext()) break

                while (iterator.hasNext()) {
                    if (!iterator.next().isAlive) iterator.remove()
                }
            }

            finished = true
            loggerInterface.accept(false, "Worker {} finished!", null, name)

            waiting.forEach { synchronized(it) { it.notify() } }
        }.start()
    }
}